---
title: "Single cell RNA-seq of cells labeled with AAV1_P5"
author: "Lukas PM Kremer"
date: "8/17/2020"
output: html_document
---

Loading required packages and defining some settings
```{r, message=FALSE, warning=FALSE, results=FALSE}
library(tidyverse)
library(here)
library(Matrix)
library(magrittr)
library(ggpointdensity)
library(patchwork)
library(DESeq2)
library(sparseMatrixStats)
library(Seurat)
```

Some aesthetic choices to determine the look of our plots
```{r}
theme_set(theme_bw() + theme(text = element_text(family = 'Arial'),
                             panel.grid.minor = element_blank(),    # remove axis grid
                             panel.grid.major = element_blank(),
                             strip.background = element_blank(),    # remove silly gray bg of facet title
                             strip.text = element_text(hjust = 0)   # left-justify facets
))

celltype_colors <- c("ENB"="#e8a623ff", "LNB"="#e17827ff", "TAP"="#20854EFF", "aNSC"="#BC3C29FF",
                     "qNSC"="#0072B5FF", "Ep"="#7876B1FF", "other"="gray50", "MG"="#6F99ADFF")
```

# Read the count matrix and corresponding metadata
```{r}
metadata <- read_tsv(here("03_single-cell-RNA-seq", "count_matrix", "metadata.tsv.gz"),
                     col_types = "ccnnncnnclnnncccnnc") %>%
  mutate(transgene_exp = fct_relevel(transgene_exp, "eYFP- NeoR-", "eYFP- NeoR+", "eYFP+ NeoR+"),
         celltype = fct_relevel(celltype, c("qNSC", "aNSC", "TAP", "ENB", "LNB", "Ep", "other")))
```

```{r}
colNames_genes <- read_tsv(here("03_single-cell-RNA-seq", "count_matrix", "colNames_genes.tsv.gz"),
                           col_names = c("gene_id_version", "gene_symbol"),
                           col_types = "cc")
symbol2ensg <- colNames_genes %>% dplyr::select(2, 1) %>% deframe()
ensg2symbol <- colNames_genes %>% deframe()
rowNames_cells <- read_tsv(here("03_single-cell-RNA-seq", "count_matrix", "rowNames_cellBarcodes.tsv.gz"),
                           col_names = "barcode", col_types = "c")

raw_counts <- readMM(here("03_single-cell-RNA-seq", "count_matrix", "raw_counts_cellXgene.mtx.gz")) %>%
  set_rownames(rowNames_cells$barcode) %>%
  set_colnames(colNames_genes$gene_id_version) %>% 
  as("dgCMatrix")

trans_counts <- log1p(1e4 * raw_counts / sparseMatrixStats::rowSums2(raw_counts))
```

A little test plot showing the number of UMIs and genes per cell
```{r}
(metadata %>%
  ggplot(aes(x = UMAP1, y = -UMAP2, color = n_genes))) +
(metadata %>%
  ggplot(aes(x = UMAP1, y = -UMAP2, color = n_UMIs))) &
  geom_point(size=.3) &
  coord_fixed() &
  scale_color_viridis_c()
```

## Figure 3b - UMAP of all cells, colored by cell type

Plot the UMAP that shows our cell types
```{r}
# define the location and text of the cell type labels to be added to the UMAP
ct_plotlabs <- tribble(
  ~celltype, ~UMAP1, ~UMAP2,
  "qNSC", 18.5, -3.7,
  "aNSC", 16, 1.5,
  "TAP", 16.5, 6.5,
  "ENB", 14.5, 10,
  "LNB", 14, 15,
  "Ep", -1.3, 6,
  "other", 5, 6,
)

metadata %>%
  ggplot(aes(y = -UMAP2, x = UMAP1, color = celltype)) +
  geom_point(size = .4, stroke = 0) +
  aes(label = celltype) +
  geom_text(data = ct_plotlabs, fontface="bold") +
  coord_fixed() +
  scale_color_manual(values = celltype_colors) +
  theme(legend.position = "none") +
  labs(x = "UMAP1", y = "UMAP2")
```

... and the small pie charts that show the proportion of off-target cells in each replicate. Since we're interested in the performance of AAV1_P5, we discard all unlabeled (eYFP negative) cells before calculating this proportion.  
In this case, off-target cells are defined as cells that are not in a cluster that is part of the main lineage (qNSC -> LNB).
```{r}
off_target <- metadata %>%
  filter(eYFP_counts > 0) %>% 
  group_by(sample, in_lineage) %>% 
  tally() %>% ungroup() %>% 
  pivot_wider(names_from = "in_lineage", values_from = "n") %>% 
  mutate(label = paste0(formatC(100 * `FALSE` / (`FALSE` + `TRUE`), format = "f", digits = 1), "%"))

off_target
```

```{r}
pie1 <- metadata %>% filter(sample == "sample #1" & eYFP_counts > 0) %>%
  left_join(off_target, by = "sample") %>%
  ggplot(aes(x = "", fill = in_lineage, label = label)) +
  geom_bar(width=1, color="white")

pie2 <- metadata %>% filter(sample == "sample #2" & eYFP_counts > 0) %>%
  left_join(off_target, by = "sample") %>%
  ggplot(aes(x = "", fill = in_lineage, label = label)) +
  geom_bar(width=1, color="white")

(pie1 + pie2) &
  scale_fill_manual(values = c("TRUE"="gray70", "FALSE"="gray50")) &
  coord_polar("y", start=0, direction=-1L) &
  theme_void() &
  theme(legend.position = "none", strip.text = element_text(hjust=1))
```

## Supplementary Figure S4h - origin of cells
```{r}
# First, calculate the 2D kernel density estimate in the UMAP for both samples
data <- tibble(x = metadata$UMAP1, y = metadata$UMAP2, group = metadata$batch)
data.group1 <- dplyr::filter(data, group == "WT1")
data.group2 <- dplyr::filter(data, group == "WT2")

adjust <- .75  # multiplicative bandwith adjustment
base.args.group1 <- list(
  x = data.group1$x,
  y = data.group1$y,
  lims = c(range(data$x), range(data$y)))
base.args.group2 <- list(
  x = data.group2$x,
  y = data.group2$y,
  lims = c(range(data$x), range(data$y)))
method.args <- list(
  n = 100,
  h = c(MASS::bandwidth.nrd(data$x), MASS::bandwidth.nrd(data$y)) * adjust
)

dens.group1 <- do.call(MASS::kde2d, c(base.args.group1, method.args))
dens.group2 <- do.call(MASS::kde2d, c(base.args.group2, method.args))

ix <- findInterval(data$x, dens.group1$x)
iy <- findInterval(data$y, dens.group1$y)
ii <- cbind(ix, iy)
data$density.group1 <- dens.group1$z[ii]
data$density.group2 <- dens.group2$z[ii]

# For every cell, subtract the two densities at this position to get the color
data %>%
  ggplot(aes(x = x, y = -y, color = density.group1 - density.group2)) +
  geom_point(size = .3) +
  scale_color_gradient2(midpoint = 0, mid = "gray70", low = "#000099", high = "#FF9900") +
  coord_fixed() +
  geom_text(data = ct_plotlabs,
            mapping = aes(x = UMAP1, y = -UMAP2, label = celltype),
            inherit.aes = F, fontface = "bold") +
  theme(legend.position = c(0.25, 0.25), legend.background = element_blank()) +
  labs(x = "UMAP1", y = "UMAP2", color = "Δ cell density")
```
```{r, include=FALSE}
# ggsave("Supplementary_Figure_S4h.svg", height = 10, width = 10, units = "cm")
```

## Supplementary Figure S4i - highlighting cells that express eYFP
```{r}
metadata %>%
  slice_sample(prop=1) %>% # shuffle point plotting order
  ggplot(aes(y = -UMAP2, x = UMAP1, color = eYFP_counts > 0)) +
  geom_point(size=.5, stroke=0) +
  geom_text(data = ct_plotlabs, fontface="bold",
            mapping = aes(y = -UMAP2, x = UMAP1, color = celltype , label = celltype)) +
  coord_fixed() +
  labs(title = "cells expressing eYFP") +
  theme(legend.position = "none") +
  scale_color_manual(values=c(celltype_colors, c("TRUE"="#ffc40c", "FALSE"="gray50"))) +
  labs(x = "UMAP1", y = "UMAP2")
```

```{r, include=FALSE}
# ggsave("Supplementary_Figure_S4i.svg", height = 10, width = 10, units = "cm")
```

## Supplementary Figure S4j- highlighting cells that express Cre
```{r}
metadata %>%
  filter(Cre_counts == 0) %>%
  ggplot(aes(y = -UMAP2, x = UMAP1)) +
  geom_point(size=.5, stroke=0, color = "gray50") +
  geom_point(size=1, stroke = 0, color="red", data = filter(metadata, Cre_counts > 0)) +
  geom_text(data = ct_plotlabs, fontface="bold",
            mapping = aes(y = -UMAP2, x = UMAP1, color = celltype , label = celltype)) +
  coord_fixed() +
  labs(title = "cells expressing Cre (plotted on top)") +
  theme(legend.position = "none") +
  scale_color_manual(values=c(celltype_colors, c("TRUE"="#ffc40c", "FALSE"="gray50"))) +
  labs(x = "UMAP1", y = "UMAP2")
```

```{r, include=FALSE}
# ggsave("Supplementary_Figure_S4j.svg", height = 10, width = 10, units = "cm")
```

## Figure 3c - proportion of cells that express eYFP or NeoR
```{r}
group_counts <- metadata %>%
  mutate(sorting = if_else(celltype %in% c("LNB", "Ep"), "eYFP+", "")) %>%
  filter(celltype != "other") %>% 
  group_by(celltype, sample, sorting) %>%
  tally()

transgene_fill <- c("eYFP- NeoR+"="#008000", "eYFP+ NeoR+"="#80A206", "eYFP- NeoR-"="gray75", "eYFP+ NeoR-"="#ffc40c")

metadata %>%
  filter(celltype != "other") %>%
  mutate(transgene_exp = fct_relevel(transgene_exp, "eYFP- NeoR-", "eYFP- NeoR+", "eYFP+ NeoR+"),
         sorting = if_else(celltype %in% c("LNB", "Ep"), "eYFP+", "")) %>%
  ggplot(aes(y = fct_rev(celltype), fill = transgene_exp)) +
  geom_bar(position="fill") +  
  geom_text(data = group_counts, size = 2,
            mapping = aes(y = fct_rev(celltype), x = .15, label = str_c("m=", n)),
            inherit.aes = F) +
  facet_grid(sorting ~ sample, space = "free", scales='free_y') +
  scale_fill_manual(values=transgene_fill) +
  scale_x_continuous(labels = scales::percent_format(accuracy = 1),
                     limits = c(0,1), breaks=0:2/2, minor_breaks = 0:4/4) +
  labs(x = "fraction of cells", y = "", fill = "") +
  theme(legend.position = "top", legend.direction = "horizontal",
        axis.text.x = element_text(angle = 45, hjust = 1)) +
  guides(fill = guide_legend(reverse = TRUE))
```

## Figure 3d - boxplots + beeswarm plots of UMIs per cell
```{r, collapse=TRUE}
metadata %>%
  mutate(sorting = if_else(celltype %in% c("LNB", "Ep"), "eYFP+", "")) %>%
  filter(celltype != "other") %>% 
  ggplot(aes(y = fct_rev(celltype), x = n_UMIs/1000, color = celltype)) +
  ggbeeswarm::geom_quasirandom(size = .25, stroke = 0, groupOnX=F) + 
  geom_boxplot(color = "black", outlier.shape = NA, fill = NA, size = .5) +
  scale_color_manual(values = celltype_colors) +
  facet_grid(sorting ~ ., scales = "free", space = "free") +
  labs(y="", x = "total UMI count\n(in thousands)") +
  theme(legend.position = "none") +
  xlim(c(NA, 50))
```

## Figure 3e - Maximum Likelihood Estimation of labeling efficiency

Add a penalty to prevent tiny mu_R (tiny expression strength of the resistance gene)  
blue = 0  
red = the minimum reasonable value for mu_R we previously observed without this penalty (sometimes it would just go to almost zero, which obviously doesn't make sense and results in very wrong predictions of almost 100% labeling efficiency).
```{r}
penalty <- function(mu) {
  1 / (1 + exp(9e5 * mu - 9))
}

tibble(mu = seq(0, 1e-4, length.out = 10000)) %>% 
  mutate(pen = penalty(mu)) %>% 
  ggplot(aes(x = mu, y = pen)) +
  geom_line() +
  geom_vline(xintercept = 2.11e-5, color = "red") +
  geom_vline(xintercept = 0, color = "blue")
```

Define the logit-transform and its inverse
```{r}
logit <- function(x) {log(x / (1 - x))}
invlogit <- function(x) {exp(x) / (1 + exp(x))}
```
Define starting points for our parameters
```{r}
starts_2genes <- c(
  logit(.5),     # p = labeling efficiency
  logit(.05),    # q = proportion of cells with incomplete Cre-mediated excision
  log(10^-3.4),  # µ_Y = overdispersion parameter of eYFP expression
  log(.2),       # alpha_Y = expression strength of eYFP (only in eYFP-expressing cells of course)
  log(10^-4.4),  # µ_R = overdispersion parameter of NeoR expression
  log(.2)        # alpha_R = expression strength of NeoR (only in NeoR-expressing cells of course)
)
```

Define our likelihood function as written in the Methods section.  
I also added some lines afterwards that make this function usable in combination with `dplyr::summarise`.
```{r}
mle_labeling_efficiency <- function(kY, kR, S, ...) {
  optim(
    starts_2genes,
    function(x) {
      p <- invlogit(x[1])
      q <- invlogit(x[2])
      muY <- exp(x[3])
      aY <- exp(x[4])
      muR <- exp(x[5])
      aR <- exp(x[6])

      -sum(log(
        p * q * dnbinom(kY, mu = muY * S, size = 1 / aY) * dnbinom(kR, mu = muR * S, size = 1 / aR) +
        ifelse(kR == 0, p * (1-q) * dnbinom(kY, mu = muY * S, size = 1 / aY), 0) +
        ifelse(kY == 0, (1-p) * dnbinom(kR, mu = muR * S, size = 1 / aR), 0) -
        penalty(muR)
      ))
    },
    ...
  ) %>% unlist() %>%
    as_tibble_row() %>%
    mutate(
      logL = -value,
      p = invlogit(par1),
      q = invlogit(par2),
      muY = exp(par3),
      aY = exp(par4),
      muR = exp(par5),
      aR = exp(par6)
    ) %>%
    dplyr::select(-starts_with("par"), -counts.function, -counts.gradient, -value)
}
```

Use the above likelihood function to estimate our parameters for every cell type per sample
```{r, message=FALSE, warning=FALSE}
mle_res <- metadata %>%
  group_by(celltype, sample) %>% 
  summarise(mle_labeling_efficiency(kY=eYFP_counts, kR=NeoR_counts, S=n_UMIs, control=list("maxit"=5000)))

mle_res
```

```{r}
mle_res_fmt <- mle_res %>%
  filter(celltype != "other") %>%
  mutate(yfp_only = p*(1-q),
         res_only = 1-p,
         both = p*q,
         sorting = if_else(celltype %in% c("LNB", "Ep"), "eYFP+", "")) %>%
  mutate(label = paste0(formatC(100 * (yfp_only + both), format = "f", digits = 0), "%"))

mle_res_fmt %>%
  pivot_longer(c("yfp_only", "res_only", "both"), names_to = "transduction", values_to = "fraction") %>%
  mutate(transduction = fct_relevel(transduction, "res_only", "both", "yfp_only")) %>%
  ggplot(aes(y = fct_rev(celltype), fill = transduction, x = fraction)) +
  geom_col() +
  geom_text(data = mle_res_fmt, size = 2,
            mapping = aes(y = fct_rev(celltype), x = yfp_only + both - .07, label = label),
            inherit.aes = F) +
  facet_grid(sorting ~ sample, scales = "free", space = "free") +
  scale_fill_manual(values=c("res_only"="#008000", "both"="#80A206", "yfp_only"="#ffc40c"),
                    labels=c("not transduced", "transduced heterozygous", "transduced")) +
  scale_x_continuous(labels = scales::percent_format(accuracy = 1),
                     limits = c(0,1), breaks=0:2/2) +
  labs(x = "estimated fraction", y = "", fill = "") +
  theme(legend.position = "top", legend.direction = "horizontal",
        axis.text.x = element_text(angle = 45, hjust = 1)) +
  guides(fill = guide_legend(reverse = TRUE))
```

## Figure 3f - Differential expression of eYFP+ vs. eYFP- cells

Mouse genes associated with GO-term [GO:0009615](https://www.ebi.ac.uk/QuickGO/term/GO:0009615) (response to virus)
```{r}
viral_resp_genes <- c('Cnot7', 'Ccl19', 'Tlr8', 'Polr3f', 'Rnasel', 'G3bp1', 'Trex1', 'Arf1', 'Gm13271', 'Htra1', 'Nlrc5', 'Src', 'Ifnlr1', 'Mill2', 'Atg7', 'Tkfc', 'Ddx58', 'Cyp1a1', 'Ifi27l2a', 'Ifi27l2b', 'Oas1g', 'Acta2', 'Aimp1', 'Ifna5', 'Il12b', 'Ifitm1', 'Exosc4', 'Il23a', 'Crcp', 'Bst2', 'Ifna16', 'Ddx41', 'Ifit1', 'Oasl2', 'Ifna12', 'Pqbp1', 'Slfn8', 'Ifngr1', 'Irf7', 'Ifit1bl1', 'Serinc5', 'Gm13290', 'Nop53', 'Rtp4', 'Mapk11', 'Riok3', 'Trim5', 'Rrp1b', 'Cxcl9', 'Tlr7', 'Ufd1', 'Skp2', 'Rsad2', 'Zmynd11', 'Gpam', 'Gli2', 'Lgals8', 'Dhx58', 'Elmod2', 'Trim6', 'Unc13d', 'Gm13276', 'Pycard', 'Gm13278', 'Irf5', 'Ticam2', 'Il33', 'Gtf2f1', 'Ifnl3', 'Pcbp2', 'Nlrp3', 'Mul1', 'Xpr1', 'Tbk1', 'Gm49380', 'Trim25', 'Fadd', 'Dtx3l', 'Gm13288', 'Polr3g', 'Trim38', 'Oas1b', 'Trim12c', 'Ifnar1', 'Eif2ak2', 'Gbp4', 'Becn1', 'Hyal3', 'Ppm1b', 'Oas2', 'Sting1', 'Ticam1', 'Lgals9', 'Tspan32', 'Tbx21', 'Ddx60', 'Dhx36', 'Rnf135', 'Mavs', 'Eif2ak4', 'Agbl5', 'Oas1e', 'Trim30a', 'Apobec3', 'Exosc5', 'Ifna9', 'Cd86', 'Il23r', 'Oas3', 'Mov10', 'Itch', 'Vwce', 'Chrm2', 'Stat1', 'Irf2', 'Sec14l1', 'Polr3c', 'Fv1', 'Rps15a', 'Trim30c', 'Cd207', 'Rnf26', 'Prf1', 'Tlr3', 'Ifnk', 'Ifnz', 'Xcl1', 'Traf3ip1', 'Oprk1', 'Gm13289', 'Agbl4', 'Gsdme', 'Ercc6', 'Ifna11', 'Ikbke', 'Uri1', 'Ddx3x', 'Nlrx1', 'Hyal1', 'Isg20', 'Bnip3', 'Ttc4', 'Ifngr2', 'Clu', 'Zmpste24', 'Ifna4', 'Cgas', 'Trim44', 'Gbf1', 'Spon2', 'Cxadr', 'Bcl2', 'Lcn2', 'Ivns1abp', 'Cd40', 'Treml4', 'Irf3', 'Mx2', 'Ap1s1', 'Duox2', 'Tlr13', 'Aicda', 'Il12rb1', 'Batf3', 'Traf3', 'Fgl2', 'Ifna13', 'Traf3ip2', 'Cfl1', 'Cd37', 'Gm13287', 'Ifna6', 'Parp9', 'Klra8', 'Il4', 'Pum2', 'Ddit4', 'Ifnab', 'Oas1f', 'Oas1a', 'Ifit3b', 'Zbp1', 'Stmn1', 'Bcl2l1', 'Setd2', 'Irak3', 'Dclk1', 'Trim12a', 'Gm13279', 'Polr3k', 'Irgm2', 'Map3k14', 'Crebzf', 'Gm13275', 'Gm45717', 'Rnase6', 'Bcl3', 'Tspan6', 'Cxcl10', 'Gm13285', 'Aim2', 'Trim11', 'Ptprc', 'Abcc9', 'Lsm14a', 'Tnfsf4', 'Ccl5', 'Ifitm6', 'Chuk', 'Tgtp1', 'Itgax', 'Ifitm2', 'Mir29b-1', 'Zdhhc1', 'Adar', 'Mmp12', 'Shfl', 'Rnf125', 'Tgtp2', 'Polr3e', 'Ankrd17', 'Card9', 'Cdk6', 'Plscr1', 'Ifng', 'Penk', 'Il2ra', 'Tarbp2', 'Nploc4', 'Irf1', 'Ilf3', 'Apob', 'Mst1r', 'Serinc3', 'Abcf3', 'Gpatch3', 'Zcchc3', 'Ifne', 'Isg15', 'Polr3b', 'F2rl1', 'Ifit1bl2', 'Myd88', 'Usp15', 'Ddx17', 'Stat2', 'Samhd1', 'Ncbp3', 'Gm13277', 'Slfn9', 'Tagap', 'Ifnb1', 'Zc3hav1', 'Polr3h', 'Ccdc130', 'Hspb1', 'Trim30d', 'Il1b', 'Hnrnpul1', 'Ptpn22', 'Tlr9', 'Sin3a', 'Mapk14', 'C1qbp', 'Ddx21', 'Il15', 'Rnf216', 'Nt5c3', 'Ddx1', 'Gm13272', 'Oasl1', 'Odc1', 'Selenok', 'Fmr1', 'Ifna7', 'Eef1g', 'Trim15', 'Ifna2', 'Pml', 'Bnip3l', 'Ifit3', 'Il12a', 'Oas1c', 'Cct5', 'Nlrp9b', 'Ifitm3', 'Hyal2', 'Ilrun', 'Gata3', 'Spn', 'Il27', 'Zc3h12a', 'Ifnar2', 'Ifitm7', 'Aen', 'Polr3d', 'Ifna14', 'Oas1h', 'Foxp3', 'Tnf', 'Ifih1', 'Wdfy4', 'Gm13283', 'Ifna15', 'Ifna1', 'Ifnl2', 'Cd8a', 'Pum1', 'Ifit2', 'Il6', 'Apobec1', 'Trim30b', 'Kcnj8', 'Oas1d')

# convert gene symbols to ensembl gene IDs
viral_resp_genes_ensg <- symbol2ensg[viral_resp_genes]

# list of transgenes (in order to remove them from the differential expression [DE] analysis)
transgenes <- c("AmpR_AAV_id.1", "Cre_AAV_id.1", "AmpR_genomic_id.1", "ERT2_genomic_id.1", "EYFP_id.1", "NeoR_KanR_id.1")
# list of lineage cell types (to exclude off-target cells from the DE analysis)
lineage_ct <- c("qNSC", "aNSC", "TAP", "ENB", "LNB")
```

Make pseudobulk samples, grouped by cell type
```{r}
pbulk_id <- metadata %>%
  filter(celltype %in% lineage_ct) %>% 
  mutate(
    transduction = if_else(eYFP_counts > 0, "T", "NT"),
    batch = str_extract(batch, "\\d"),
    pbulk = as.character(glue::glue("{celltype}_{transduction}_{batch}"))) %>% 
  dplyr::select(bc=barcode, i=pbulk)

pbulk_groups <- pbulk_id$i %>% unique() %>% sort()

pseudobulk <- sapply(pbulk_groups, function(.x) {
  barcodes <- filter(pbulk_id, i == .x)$bc
  sparseMatrixStats::colSums2(raw_counts[barcodes, ])
}) %>%
  magrittr::set_rownames(colnames(raw_counts))

head(pseudobulk)
```

Run DESeq2 on pseudobulk samples. We're using the likelihood ratio test (LRT) to test the
full model `~ batch + cluster + condition` against the reduced model `~ batch + cluster`.
In this case, `condition` is a boolean variable indicating whether cells express eYFP or not.
```{r, collapse=TRUE}
condition <- factor(str_extract(colnames(pseudobulk), "_N?T_"))
col_data <- DataFrame(
  condition = condition,
  batch = factor(str_extract(colnames(pseudobulk), "\\d")),
  cluster = factor(str_extract(colnames(pseudobulk), "[A-Za-z]+"))
) %>% set_rownames(colnames(pseudobulk))

dds <- DESeqDataSetFromMatrix(
  pseudobulk[!(colnames(raw_counts) %in% transgenes), ],
  col_data,
  ~ batch + cluster + condition
)

dds <- DESeq(dds, test="LRT", reduced = ~ batch + cluster)
res <- results(dds) %>%
  as_tibble(rownames = "gene")
```

There are the following 18 differentially expressed genes:
```{r}
res %>% filter(padj <= 0.05) %>% arrange(padj) %>% mutate(symbol = ensg2symbol[gene]) %>% dplyr::select(-stat)
```

Plot the results (Fig 3f)
```{r, collapse=T}
p_ma <- res %>%
  mutate(is_viral = gene %in% viral_resp_genes_ensg) %>%
  arrange(is_viral) %>%
  ggplot(aes(x = baseMean, y = log2FoldChange, color = is_viral, size = is_viral)) +
  geom_point(stroke=0) +
  scale_x_log10(label=scales::label_number(drop0trailing=T)) +
  scale_color_manual(values = c("TRUE"="#051094", "FALSE"="gray60")) +
  scale_size_manual(values = c("TRUE"=.9, "FALSE"=.33)) +
  labs(x = "mean of normalized counts", y = "log2 fold change") +
  theme(legend.position = "none") +
  annotate("text", x = 100, y = 2.6, label = "viral response genes", color = "#051094")

p_dens <- res %>%
  filter(!(gene %in% transgenes)) %>%
  ggplot(aes(y = log2FoldChange)) +
  geom_density(color = "gray60", fill = "gray60", adjust = 2) +
  geom_density(data = filter(res, gene %in% viral_resp_genes_ensg),
               mapping = aes(y = log2FoldChange),
               color = "#1338BE", fill = NA, adjust = 2,
               inherit.aes = F, trim = T) +
  scale_x_continuous(breaks = c(0, 1)) +
  labs(x = "density", y = "")

p_ma + p_dens &
  ylim(-4, 4) &
  plot_layout(widths = c(5, 1))
```

```{r, include=FALSE}
# ggsave("Figure_3f.svg", height = 8, width = 10, units = "cm")
```

# Supplementary Figure S4k - Differential expression of eYFP+ vs. eYFP-/NeoR+ cells

Here, we consider eYFP+ cells as transduced and eYFP-/NeoR+ as non-transduced.  
In this comparison we have very few cells because NeoR+ cells are quite rare. Hence we cannot include cell type as a covariate.
```{r, collapse=T}
# get the indices of cells of the four different pseudobulk groups
transduced_bc_1 <- metadata %>%
  filter(eYFP_counts > 0 & batch == "WT1" & celltype %in% lineage_ct) %>%
  pull(barcode)
transduced_bc_2 <- metadata %>%
  filter(eYFP_counts > 0 & batch == "WT2" & celltype %in% lineage_ct) %>%
  pull(barcode)
not_transduced_bc_1 <- metadata %>%
  filter(eYFP_counts == 0 & NeoR_counts > 0 & batch == "WT1" & celltype %in% lineage_ct) %>%
  pull(barcode)
not_transduced_bc_2 <- metadata %>%
  filter(eYFP_counts == 0 & NeoR_counts > 0 & batch == "WT2" & celltype %in% lineage_ct) %>%
  pull(barcode)

# aggregate single cell counts to pseudobulk
pseudobulk <- cbind(
  sparseMatrixStats::colSums2(raw_counts[transduced_bc_1, ]),
  sparseMatrixStats::colSums2(raw_counts[transduced_bc_2, ]),
  sparseMatrixStats::colSums2(raw_counts[not_transduced_bc_1, ]),
  sparseMatrixStats::colSums2(raw_counts[not_transduced_bc_2, ])
) %>%
  magrittr::set_rownames(colnames(raw_counts)) %>%
  magrittr::set_colnames(c("sample1_transduced", "sample2_transduced", "sample1_rest", "sample2_rest"))

# run DESeq2
condition <- factor(c("transduced", "transduced", "rest", "rest"))
dds <- DESeqDataSetFromMatrix(
  pseudobulk,
  DataFrame(condition),
  ~ condition
)

dds <- DESeq(dds)
res <- results(dds) %>%
  as_tibble(rownames = "gene")
```


```{r, collapse=T}
sig_res <- res %>% filter(padj <= 0.05) %>% 
  mutate(symbol = ensg2symbol[gene])

p_ma <- res %>%
  filter(!(gene %in% transgenes)) %>% 
  mutate(is_viral = gene %in% viral_resp_genes_ensg) %>%
  arrange(is_viral) %>% 
  ggplot(aes(x = baseMean, y = log2FoldChange, color = is_viral, size = is_viral)) +
  geom_point(stroke=0) +
  scale_x_log10(label=scales::label_number(drop0trailing=T)) +
  scale_color_manual(values = c("TRUE"="#1338BE", "FALSE"="gray60")) +
  scale_size_manual(values = c("TRUE"=.9, "FALSE"=.33)) +
  labs(x = "mean of normalized counts", y = "log2 fold change") +
  theme(legend.position = "none") +
  annotate("text", x = 10, y = -5, label = "viral response genes", color = "#1338BE")

p_dens <- res %>%
  filter(!(gene %in% transgenes)) %>%
  ggplot(aes(y = log2FoldChange)) +
  geom_density(color = "gray60", fill = "gray60", adjust = 2) +
  geom_density(data = filter(res, gene %in% viral_resp_genes_ensg),
               mapping = aes(y = log2FoldChange),
               color = "#1338BE", fill = NA, adjust = 2,
               inherit.aes = F, trim = T) +
  labs(y = "")

p_ma + p_dens &
  ylim(-6.6, 6.6) &
  plot_layout(widths = c(5, 1))
```

```{r, include=FALSE}
# ggsave("Supplementary_Figure_S4k.svg", height = 8, width = 10, units = "cm")
```
High log2foldChange = higher expression in transduced cells  
Low log2foldChange = lower expression in transduced cells  

# Supplementary Figure S4l - eYFP+ cell vs. v-SVZ cells from [Kalamakis et al. 2019](https://pubmed.ncbi.nlm.nih.gov/30827680/)

Differential expression of our eYFP+ cells vs. adult NSC-lineage cells from the v-SVZ from Kalamakis et al. 2019

***

First, integrate the two datasets into one using the SCTransform-workflow described in the Seurat vignette [here](https://satijalab.org/seurat/v3.2/integration.html). The aim is to remove batch effects in order to make these two different datasets roughly comparable.
```{r, message=FALSE, warning=FALSE}
kalamakis_data <- Read10X(data.dir = here("03_single-cell-RNA-seq", "external_data", "Kalamakis_2019"))
metadata_kalamakis <- read.csv(here("03_single-cell-RNA-seq", "external_data", "Kalamakis_2019", "metadata_kalamakis.csv"), row.names = 1)
kalamakis <- CreateSeuratObject(counts = kalamakis_data, project = "kalamakis", meta.data = metadata_kalamakis)

# Prepare our data to use gene symbols too, instead of Ensembl IDs
colname_helper <- colNames_genes %>%
  group_by(gene_id_version) %>%
  mutate(n = n()) %>% ungroup() %>%
  mutate(keep = !is.na(gene_symbol) & n == 1 & !str_detect(gene_symbol, "_id$"))

# Create a Seurat object from our own eYFP+ cells
aavtrace <- raw_counts[raw_counts[ , "EYFP_id.1"] > 0, colname_helper %>% pull(keep)] %>%
  magrittr::set_colnames(colname_helper %>% filter(keep) %>% pull(gene_symbol)) %>%
  t() %>%
  CreateSeuratObject(project="aavtrace", assay="RNA")

aavtrace
```

Little helper function to prevent the annoying `SCTransform` function from spamming so much...
```{r}
quiet <- function(x) {
  sink(tempfile())
  on.exit(sink())
  invisible(force(x))
}
```

Integrate the datasets as shown in the vignette
```{r, message=FALSE, warning=FALSE, results=FALSE}
samples <- list("kalamakis" = kalamakis, "aavtrace" = aavtrace)
for (i in names(samples)) {
    samples[[i]] <- quiet(SCTransform(samples[[i]], verbose = FALSE))
}
```

```{r, collapse=TRUE}
int.features <- SelectIntegrationFeatures(object.list = samples, nfeatures = 1500)
int.list <- PrepSCTIntegration(object.list = samples, anchor.features = int.features)

# specify our data as the reference dataset
reference_dataset <- which(names(samples) == "aavtrace")

int.anchors <- FindIntegrationAnchors(object.list = int.list, normalization.method = "SCT", 
    anchor.features = int.features, reference = reference_dataset)
int.integrated <- IntegrateData(anchorset = int.anchors, normalization.method = "SCT")

int.integrated <- RunPCA(object = int.integrated, verbose = FALSE)
int.integrated <- RunUMAP(object = int.integrated, dims = 1:35, n.neighbors = 20L)

DimPlot(int.integrated)
```

Calculate the neighbor graph on the integrated dataset and cluster cells
```{r, collapse=TRUE}
int.integrated <- FindNeighbors(int.integrated, dims = 1:35)
int.integrated <- FindClusters(int.integrated, resolution = 0.1, algorithm = 4)
```

Plot the two integrated datasets colored by source
```{r}
int_tbl <- tibble(
  bc = colnames(int.integrated),
  source = FetchData(int.integrated, "orig.ident")[ , 1],
  UMAP_1 = FetchData(int.integrated, "UMAP_1")[ , 1],
  UMAP_2 = FetchData(int.integrated, "UMAP_2")[ , 1],
  cluster = FetchData(int.integrated, "seurat_clusters")[ , 1]
) %>%
  mutate(replicate = str_extract(bc, "\\d")) %>%
  filter(cluster %in% c(1, 2, 3, 4) & (UMAP_2 < 5 | UMAP_1 < 5))

int_tbl %>% 
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = cluster)) +
  geom_point(size=.4) +
  coord_fixed()
```

Aggregate single cell transcriptomes to pseudobulk
```{r}
pbulk_id <- int_tbl %>%
  mutate(
    pbulk = as.character(glue::glue("{source}_rep{replicate}_c{cluster}"))) %>% 
  dplyr::select(bc, i=pbulk)

pbulk_groups <- pbulk_id$i %>% unique() %>% sort()

pseudobulk <- sapply(pbulk_groups, function(.x) {
  barcodes <- filter(pbulk_id, i == .x)$bc
  sparseMatrixStats::rowSums2(int.integrated@assays$RNA@counts[ , barcodes])
}) %>%
  magrittr::set_rownames(rownames(int.integrated@assays$RNA@counts))

head(pseudobulk)
```

Run DESeq2 between the two data sets. Here we have enough cells to add celltype as a covariate again
```{r, collapse=TRUE}
condition <- factor(str_extract(colnames(pseudobulk), "[a-z]+"))
col_data <- DataFrame(
  condition = condition,
  batch = factor(str_extract(colnames(pseudobulk), "rep\\d")),
  cluster = factor(str_extract(colnames(pseudobulk), "c\\d"))
) %>% set_rownames(colnames(pseudobulk))

dds <- DESeqDataSetFromMatrix(
  pseudobulk,
  col_data,
  ~ batch + cluster + condition
)

dds <- DESeq(dds, test="LRT", reduced = ~ batch + cluster)
res <- results(dds) %>%
  as_tibble(rownames = "gene")
```

Plot the MA-plot shown in Supplementary Figure S4l
```{r, collapse=TRUE}
p_ma <- res %>%
  mutate(is_viral = gene %in% viral_resp_genes) %>%
  arrange(is_viral) %>% 
  ggplot(aes(x = baseMean, y = log2FoldChange, color = is_viral, size = is_viral)) +
  geom_point(stroke=0) +
  scale_x_log10(label=scales::label_number(drop0trailing=T)) +
  scale_color_manual(values = c("TRUE"="#1338BE", "FALSE"="gray60")) +
  scale_size_manual(values = c("TRUE"=.9, "FALSE"=.33)) +
  labs(x = "mean of normalized counts", y = "log2 fold change") +
  theme(legend.position = "none") +
  annotate("text", x = 10, y = -11, label = "viral response genes", color = "#1338BE")

p_dens <- res %>%
  filter(!(gene %in% transgenes)) %>%
  ggplot(aes(y = log2FoldChange)) +
  geom_density(color = "gray60", fill = "gray60", adjust = 2) +
  geom_density(data = filter(res, gene %in% viral_resp_genes),
               mapping = aes(y = log2FoldChange),
               color = "#1338BE", fill = NA, adjust = 2,
               inherit.aes = F, trim = T) +
  scale_x_continuous(breaks = c(0, 1)) +
  labs(x = "density", y = "")

p_ma + p_dens &
  ylim(-15, 15) &
  plot_layout(widths = c(5, 1))
```

```{r, include=FALSE}
# ggsave("Supplementary_Figure_S4l.svg", height = 8, width = 10, units = "cm")
```

### sessionInfo()
```{r}
sessionInfo()
```
